### PR TITLE
Health optimisation

### PR DESCRIPTION
**Pull request by Bod9001**
Could do with more work but
good for early review

fixes reagent SO ID

CL: [Fix] comprehensive health optimizations (should improve server fps significantly).

### CODE REVIEW

File: UnityProject/Assets/Scripts/Systems/Chemistry/Reaction.cs [hunk 1]
```diff
@@ -92,6 +93,11 @@ public bool HasIngredients(ReagentMix reagentMix)
 				return false;
 			}
 
+			if (ingredients.m_dict.Count > reagentMix.reagents.m_dict.Count)
```
Comment (VernaSaurus) [inline UnityProject/Assets/Scripts/Systems/Chemistry/Reaction.cs:96]: what does this do? does this cause a reaction to fail if there are not enough reagents to even bother checking if it should react?

File: UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs [hunk 2]
```diff
@@ -123,6 +124,12 @@ public static void Reset()
 		AssetDatabase.SaveAssets();
 	}
 
+
+	public static double GetRandomNumberInRange(double minNumber, double maxNumber)
+	{
+		return new System.Random().NextDouble() * (maxNumber - minNumber) + minNumber;
+	}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs]: wut, that does not do what the function is called

Comment (corp-0) [reply UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs]: is this for generating IDs again? Can't you just use C#'s UUID generator?

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs]: this is for just testing

File: UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs [hunk 3]
```diff
@@ -133,6 +140,109 @@ public static void Generate()
...
+		ST.Start();
+
+		var List = Required.Intersect(Container);
+
+		if (List.Count() == Required.Count)
+		{
+		}
+		else
+		{
+		}
```
Comment (Aranclanos) [inline UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs]: ?

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/Editor/GenerateSpriteSO.cs]: this is for just testing

**Review by ThatDan123 — DISMISSED** [review]

**Comment by ThatDan123** [comment]
oh, tests need fixing here too

File: UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs [hunk 4]
```diff
@@ -10,13 +11,13 @@ namespace ScriptableObjects
 	public class ChemistryReagentsSO : SingletonScriptableObject<ChemistryReagentsSO>
 	{
 		[SerializeField]
-		private Reagent[]  allChemistryReagents = new Reagent[0];
+		private List<Reagent> allChemistryReagents = new List<Reagent>();
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs:14]: why change this to list?

File: UnityProject/Assets/Scripts/Items/Implants/Organs/Liver.cs [hunk 5]
```diff
@@ -85,16 +85,16 @@ private void BloodToLiver(ReagentContainerBody blood)
 			//figure out how much we are going to process or remove
 			lock (blood.CurrentReagentMix.reagents)
 			{
-				foreach (Reagent reagent in blood.CurrentReagentMix.reagents.Keys)
+				foreach (var reagent in blood.CurrentReagentMix.reagents.m_dict)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Implants/Organs/Liver.cs]: what's the ``m_dict``?

File: UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs:14]: I think a little bit ago there was bugs with SOs and arrays but might be fixed now, doesn't really make any difference

File: UnityProject/Assets/Scripts/Items/Implants/Organs/Liver.cs (hunk 5 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Items/Implants/Organs/Liver.cs]: it's the dictionary at the base of the serialisable dictionary

Comment (Aranclanos) [reply UnityProject/Assets/Scripts/Items/Implants/Organs/Liver.cs]: serializable dictionaries are really bad with performance and gc, accessing it directly makes it way better in that regard, (m_dict was private previously)
I had a plan to remove them eventually

File: UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs (hunk 4 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs:14]: I thought lists were more expensive than arrrays

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/ScriptableObjects/ChemistryReagentsSO.cs:14]: https://stackoverflow.com/questions/454916/performance-of-arrays-vs-lists

**Review by NoooneyDude — APPROVED** [review]

**Review by ThatDan123 — APPROVED** [review]

**Pull request merged**